src/data/nat/choose.lean

```diff
@@ -100,8 +102,76 @@ theorem add_pow [comm_semiring α] (x y : α) (n : ℕ) :
+  ∑ j in (Ico (m + 1 - i) (m + 1)), f j = ∑ j in Ico (m + 1) (m + 1 + i), f j :=
+begin
+  induction i with i,
+  { simp },
+  { have t : (m + 1) + (i + 1) = ((m + 1) + i) + 1, norm_num,
+    rw t, clear t,
+    rw ← (@sum_Ico_succ_top _ _ (m + 1) ((m + 1) + i) (by exact nat.le.intro rfl) f).symm,
+    rw ← i_ih, clear i_ih,
+    have t : f (m + 1 + i) = f (m - i), by
+      { have munge : m + 1 + i ≤ 2 * m + 1,
+          { calc m + 1 + i ≤ m + 1 + m : by exact add_le_add_left (nat.lt_succ_iff.mp i_bound) (m + 1)
+          ... = 2 * m + 1 : by ring_exp, },
+        have v : 2 * m + 1 - (m + 1 + i) = m - i,
+          { calc 2 * m + 1 - (m + 1 + i)
+              = 2 * m + 1 - (m + 1) - i : eq.symm (nat.sub_sub (2 * m + 1) (m + 1) i)
+             ... = m + m + 1 - m.succ - i : by rw ← two_mul m
+             ... = m + (m + 1) - (m + 1) - i : by rw ← add_assoc m m 1
+             ... = m - i : by rw nat.add_sub_cancel m m.succ, },
+        have reflected : f (m + 1 + i) = f (2 * m + 1 - (m + 1 + i)), { exact reflects (m + 1 + i) munge },
+        rw v at reflected,
+        exact reflected, },
+    rw t, clear t,
+    { rw nat.succ_sub (nat.lt_succ_iff.mp i_bound),
+      have s : f (m - i) + ∑ j in (Ico (m - i + 1) (m + 1)), f j = ∑ j in (Ico (m - i) (m + 1)), f j,
+        { exact (@sum_eq_sum_Ico_succ_bot _ _ (m - i) (m + 1) (nat.sub_lt_succ m i) f).symm, },
+      simp only [succ_sub_succ_eq_sub],
+      rw ← s,
+      ring, },
+    exact le_of_lt i_bound, }
+end
+
+private lemma sum_range_reflects_halfway (m : nat) (f : nat → nat)
+  (reflects : ∀ x ≤ 2 * m + 1, f x = f (2 * m + 1 - x)) :
+  ∑ i in (range (m + 1)), f i = ∑ i in (Ico (m + 1) (2 * m + 2)), f i :=
+begin
+  have r : 2 * m + 2 = 2 * (m + 1), ring,
+  rw r,
+  simpa [finset.Ico.zero_bot (m + 1), two_mul (m + 1)] using (reflect_sum_lemma (m + 1) m (le_refl _) f reflects),
+end
+
+lemma sum_range_choose_halfway (m : nat) :
+  ∑ i in range (m + 1), nat.choose (2 * m + 1) i = 4 ^ m :=
+begin
+  have reflects : ∀ x ≤ 2 * m + 1, choose (2 * m + 1) x = choose (2 * m + 1) (2 * m + 1 - x),
+  { intros x pr,
+    exact eq.symm (@choose_symm (2 * m + 1) x pr), },
+
+  have v : 2 * (∑ i in range (m + 1), nat.choose (2 * m + 1) i) = 2 * 4 ^ m,
+    calc 2 * (∑ i in range (m + 1), nat.choose (2 * m + 1) i)
+          = ∑ i in range (m + 1), nat.choose (2 * m + 1) i + ∑ i in range (m + 1), nat.choose (2 * m + 1) i : by ring
+      ... = ∑ i in range (m + 1), nat.choose (2 * m + 1) i + ∑ i in Ico (m + 1) (2 * m + 2), nat.choose (2 * m + 1) i
+              : by rw (sum_range_reflects_halfway m (choose (2 * m + 1)) reflects)
+      ... = ∑ i in range (2 * m + 2), nat.choose (2 * m + 1) i
+            : by rw @sum_range_add_sum_Ico _ _ (choose (2 * m + 1)) (m + 1) (2 * m + 2) (by linarith)
+      ... = 2 ^ (2 * m + 1) : sum_range_choose (2 * m + 1)
+      ... = 2 * 2 ^ (2 * m) : by ring_exp
+      ... = 2 * 4 ^ m : by { rw nat.pow_mul 2 m 2, refl, },
+
```

Comment:
```suggestion
```

Reply:
I do generally like newlines around long statements, because it means I can vim-jump more easily (with `{` and `}`).